### PR TITLE
Guard against -1 Returned from sysconf for the Cache Sizes Causing Large Gen0 Sizes and Budgets for Certain Linux Distributions.

### PR DESCRIPTION
**Pull request by mrsharm**
## Issue Summary

- ``GetLogicalProcessorCacheSizeFromOS`` retrieves a bogus value of ``SIZE_MAX`` (-1 but type casted to a ``size_t``) from the ``sysconf`` call for the L4 Cache Size that's now recognized has the last cache size (as it's not 0) causing a significantly larger gen0 size and budget for Debian 12 vs. Debian 11 for the same hardware i.e., same cache sizes, architecture etc. This behavior also seems to show up for Ubuntu 22.04.
- This is confirmed by running ``getconf -a | grep LEVEL4_CACHE`` on a Debian 12 (and an Ubuntu 22.04) machine results in an empty string for the L4 cache size - this is different from the behavior of Debian 11 that returns a 0 for the L4 cache size. 
- ``SIZE_MAX`` from ``GetLogicalProcessorCacheSizeFromOS`` is also returned from ``GetCacheSizePerLogicalCpu`` and causes an extremely large gen0 size thereby affecting the gen0 budget for Server GC. 

### Further Details

Running ``getconf -a | grep LEVEL4_CACHE`` Outputs:

#### Debian 11

LEVEL4_CACHE_SIZE                  0
LEVEL4_CACHE_ASSOC                 0
LEVEL4_CACHE_LINESIZE              0

#### Debian 12

LEVEL4_CACHE_SIZE
LEVEL4_CACHE_ASSOC
LEVEL4_CACHE_LINESIZE

<hr>

#### Ubuntu 20.04

LEVEL4_CACHE_SIZE                  0
LEVEL4_CACHE_ASSOC                 0
LEVEL4_CACHE_LINESIZE              0

#### Ubuntu 22.04

LEVEL4_CACHE_SIZE
LEVEL4_CACHE_ASSOC
LEVEL4_CACHE_LINESIZE

#### How To Detect This Issue

Run ``getconf -a | grep LEVEL4_CACHE`` and observe if the value for the LEVEL4_CACHE is empty or 0. If it is empty, it indicates that sysconf will return -1 and therefore, we will be susceptible to a large gen0 size and not if it's 0.

## Solution

- Guard against ``-1`` from being returned from sysconf that we cast to a size_t and therefore becomes ``SIZE_MAX`` to prevent much larger gen0 sizes and budgets than if -1 wasn't returned from sysconf.

## Testing

First, we confirmed that running a standalone C program on Debian 11 vs. Debian 12 with code:

```c
#include <unistd.h>
#include <stdio.h>

int main (int argc, char **argv) {
    printf( " Raw LEVEL4 Cache Size: %lu \n",  sysconf( _SC_LEVEL4_CACHE_SIZE ));
    return 0;
}
```

Results in:

Debian 11:
Raw LEVEL4 Cache Size: 0
 
Debian 12:
Raw LEVEL4 Cache Size: 18446744073709551615 or SIZE_MAX

Ubuntu 22.04:
Raw L4 Cache Value: 18446744073709551615 or SIZE_MAX

Additionally, with this fix, the customer confirmed that this fix resulted in their memory going back to previous levels.

### CODE REVIEW

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 1]
```diff
@@ -868,7 +868,7 @@ bool ReadMemoryValueFromFile(const char* filename, uint64_t* val)
     return result;
 }
 
-#define UPDATE_CACHE_SIZE_AND_LEVEL(NEW_CACHE_SIZE, NEW_CACHE_LEVEL) if (NEW_CACHE_SIZE > cacheSize) { cacheSize = NEW_CACHE_SIZE; cacheLevel = NEW_CACHE_LEVEL; }
+#define UPDATE_CACHE_SIZE_AND_LEVEL(NEW_CACHE_SIZE, NEW_CACHE_LEVEL) if ((NEW_CACHE_SIZE != SIZE_MAX) && (NEW_CACHE_SIZE > cacheSize)) { cacheSize = NEW_CACHE_SIZE; cacheLevel = NEW_CACHE_LEVEL; }
```
Comment (jkotas) [inline src/coreclr/gc/unix/gcenv.unix.cpp:871]: Is the problem that some of the system APIs return -1 when the cache does not exist?

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 2]
```diff
@@ -894,7 +894,7 @@ static size_t GetLogicalProcessorCacheSizeFromOS()
 #endif
 
 #if defined(TARGET_LINUX) && !defined(HOST_ARM) && !defined(HOST_X86)
-    if (cacheSize == 0)
+    if (cacheSize == 0 || cacheSize == SIZE_MAX)
```
Comment (Maoni0) [inline src/coreclr/gc/unix/gcenv.unix.cpp:897]: since you are already filtering out `SIZE_MAX` in the `UPDATE_CACHE_SIZE_AND_LEVEL` macro, how can cacheSize be `SIZE_MAX`?

**Comment by Maoni0** [comment]
>                     cacheSize = std::max(cacheSize, size);

what if size is `SIZE_MAX` here?

---
Refers to: src/coreclr/gc/unix/gcenv.unix.cpp:925 in 5868885. [](commit_id = 5868885759075c7843e55a5acb01db5c52440a4b, deletion_comment = False)

File: src/coreclr/gc/unix/gcenv.unix.cpp (hunk 1 above)
Comment (mrsharm) [reply src/coreclr/gc/unix/gcenv.unix.cpp:871]: We found an issue where sysconf is returning SIZE_MAX for the L4 Cache in a few Linux OSes. Have updated the PR description for all details.

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 3]
```diff
@@ -912,7 +912,9 @@ static size_t GetLogicalProcessorCacheSizeFromOS()
         {
             path_to_size_file[index] = (char)(48 + i);
 
-            if (ReadMemoryValueFromFile(path_to_size_file, &size))
+            // Only accept reading cache sizes from size files if they are 
+            // non-bogus values i.e., non-zero and != SIZE_MAX.
+            if (ReadMemoryValueFromFile(path_to_size_file, &size) && ((size != SIZE_MAX) || (size != 0)))
```
Comment (janvorli) [inline src/coreclr/gc/unix/gcenv.unix.cpp:917]: Do we need to keep checking for the size at all the places? Wouldn't it be sufficient to do it in the macro? It doesn't seem the `ReadMemoryValueFromFile` would ever return the `size` equal to `SIZE_MAX` and if it returned 0, it should not cause any trouble.

**Comment by mrsharm** [comment]
> > ```
> >                 cacheSize = std::max(cacheSize, size);
> > ```
> 
> what if size is `SIZE_MAX` here?
> 
> Refers to: src/coreclr/gc/unix/gcenv.unix.cpp:925 in [5868885](https://github.com/dotnet/runtime/commit/5868885759075c7843e55a5acb01db5c52440a4b). [](commit_id = [5868885](https://github.com/dotnet/runtime/commit/5868885759075c7843e55a5acb01db5c52440a4b), deletion_comment = False)

Guarding against this now by rejecting any 0 or SIZE_MAX values here.

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 4]
```diff
@@ -912,19 +914,24 @@ static size_t GetLogicalProcessorCacheSizeFromOS()
...
 
                 if (ReadMemoryValueFromFile(path_to_level_file, &level))
                 {
                     UPDATE_CACHE_SIZE_AND_LEVEL(size, level)
                 }
-                else
+
+                // We guard against the case where the size = -1 or 0. 
+                // This isn't currently an issue but to be safe, we add this check.
+                else if (size > 0)
```
Comment (janvorli) [inline src/coreclr/gc/unix/gcenv.unix.cpp:932]: I don't think we need this guard here. 0 size would be ok and the first `ReadMemoryValueFromFile` should fail when it cannot extract a size.

Comment (mrsharm) [reply src/coreclr/gc/unix/gcenv.unix.cpp:932]: CC: @Maoni0 - we were discussing that this was a redundant check, but it'll be still good to add to prevent issues where the underlying mechanism to get cache size information changes. Such as in this case where we didn't know that the sysconf mechanism was different. 

I don't feel too strongly about this - please let me know what you think, and I can change this code accordingly.

Comment (Maoni0) [reply src/coreclr/gc/unix/gcenv.unix.cpp:932]: right, it would guard against both 0 and -1. also the comment really should just be deleted.

looking at this function, it seems like we have some inconsistency between this for loop and the check below that says `if (cacheLevel != 3)`, presumably the for loop is saying we can get a level 4 cache that is valid yet it will be overwritten by the `if (cacheLevel != 3)` code below since it's not 3. but that's a different issue.

**Review by jkotas — APPROVED** [review]
Looks good to me as a minimal delta fix for servicing backport.

(I do not particularly like how some of the local variables are used in the macro, but others are not. It makes the code hard to follow. Cleaning it up is not necessary for the servicing fix.)

**Comment by janvorli** [comment]
> I do not particularly like how some of the local variables are used in the macro, but others are not. It makes the code hard to follow.

I completely agree that it would be good to eventually clean this up. Macros using ambient stuff are always difficult to follow, debug and reason about.

**Comment by debracey** [comment]
Possibly related to 

https://github.com/dotnet/aspnetcore/issues/54405
